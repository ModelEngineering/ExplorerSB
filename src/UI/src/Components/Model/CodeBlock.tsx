import { useState } from "react";
import SyntaxHighlighter from "react-syntax-highlighter";
import { vscDarkPlus as stylesheet } from "react-syntax-highlighter/dist/esm/styles/prism";
import { BiCopy as CopyIcon, BiCheck as CheckIcon } from "react-icons/bi";
import { CopyToClipboard } from "react-copy-to-clipboard";

// The CodeBlock component is used to display code
// It uses the react-syntax-highlighter library to display the code
// It uses the react-copy-to-clipboard library to allow the user to copy the code to their clipboard
// It takes in the code string as a prop
// It renders the code string in a code block
// It also allows the user to copy the code to their clipboard with a button
// It displays a check mark for 2.5 seconds when the code is copied

const CodeBlock = ({ codeString }: { codeString: string }) => {
  const [copied, setCopied] = useState(false);
  const copy = () => {
    setCopied(true);
    setTimeout(() => {
      setCopied(false);
    }, 2500);
  };
  return (
    <div style={{position: "relative"}}>
      <CopyToClipboard text={codeString} onCopy={copy}>
        <button className="copy-button">{copied ? <CheckIcon /> : <CopyIcon />}</button>
      </CopyToClipboard>
      <SyntaxHighlighter
        language="markup"
        style={stylesheet}
        showLineNumbers={true}
        customStyle={{margin: "0"}}
      >
        {codeString}
      </SyntaxHighlighter>
    </div>
  );
};

export default CodeBlock;
