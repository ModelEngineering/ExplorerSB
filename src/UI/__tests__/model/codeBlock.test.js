import React from "react";
import { render } from "@testing-library/react";
import CodeBlock from "../../components/model/codeBlock";

describe("CodeBlock component", () => {
  it("renders correctly", () => {
    const codeString = `const greeting = "Hello, world!";\nconsole.log(greeting);`;
    const { container } = render(<CodeBlock codeString={codeString} />);
    expect(container).toMatchSnapshot();
  });
});
