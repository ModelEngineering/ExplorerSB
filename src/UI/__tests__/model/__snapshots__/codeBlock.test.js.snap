// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CodeBlock component renders correctly 1`] = `
<div>
  <div
    class="relative"
  >
    <button
      class="absolute right-0 top-0 m-4 hidden rounded-md border-2 border-slate-700 bg-slate-700 bg-opacity-50 px-4 py-2 text-white hover:border-slate-600 hover:bg-opacity-70 group-hover:block"
    >
      <svg
        fill="currentColor"
        height="1em"
        stroke="currentColor"
        stroke-width="0"
        viewBox="0 0 24 24"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M20 2H10c-1.103 0-2 .897-2 2v4H4c-1.103 0-2 .897-2 2v10c0 1.103.897 2 2 2h10c1.103 0 2-.897 2-2v-4h4c1.103 0 2-.897 2-2V4c0-1.103-.897-2-2-2zM4 20V10h10l.002 10H4zm16-6h-4v-4c0-1.103-.897-2-2-2h-4V4h10v10z"
        />
      </svg>
    </button>
    <pre
      style="color: rgb(212, 212, 212); font-size: 13px; text-shadow: none; font-family: Menlo, Monaco, Consolas, "Andale Mono", "Ubuntu Mono", "Courier New", monospace; direction: ltr; text-align: left; white-space: pre; word-spacing: normal; word-break: normal; line-height: 1.5; tab-size: 4; hyphens: none; padding: 1em; margin: 0px; overflow: auto; background: rgb(30, 30, 30);"
    >
      <code
        class="className"
        style="white-space: pre;"
      >
        <span
          class="linenumber react-syntax-highlighter-line-number"
          style="display: inline-block; min-width: 1.25em; padding-right: 1em; text-align: right; user-select: none; color: rgb(106, 153, 85);"
        >
          1
        </span>
        <span>
          const 
        </span>
        <span
          class="hljs-attr"
        >
          greeting
        </span>
        <span>
           = 
        </span>
        <span
          class="hljs-string"
        >
          "Hello, world!"
        </span>
        <span
          class="hljs-comment"
        >
          ;
        </span>
        <span>
          

        </span>
        <span
          class="linenumber react-syntax-highlighter-line-number"
          style="display: inline-block; min-width: 1.25em; padding-right: 1em; text-align: right; user-select: none; color: rgb(106, 153, 85);"
        >
          2
        </span>
        <span>
          console.log(greeting)
        </span>
        <span
          class="hljs-comment"
        >
          ;
        </span>
      </code>
    </pre>
  </div>
</div>
`;
